### ISSUE TITLE
Support dynamic tag names in template tags containing HTML (e.g. lit-html)

### ISSUE THREAD

**Issue by bashmish** [prettier/prettier#5922]
**Prettier 1.16.4**
[Playground link](https://prettier.io/playground/#N4Igxg9gdgLgprEAuc0DOMAEMCGBzTAXkwHIATASwDcSBuAHSkig2zgFsAHAGx3iMwALGO24ADRgB4AJMFx4AvgD4pAelnzljMQyggANCAicYFdMlA4ATlYgB3AArWEaZCBxUIFMgZAAjKxwwAGs4GABlTiCKKDxkGCsAVzhDYVEAdUEKeDQosDhwl2zqbIBPNzA0V0MYtDgrGAdAvHYcZAAzHG46wwArNAAPACFAkLDwnHY4ABkYuA6unpB+gfCYvG44AEVEiHgF7pSQKKs6qzc-HD9S7mhfTisYmHTvGEFkAA4ABkMHiDr0oFOG4HnAzlR5oYrHAAI6JCjQpr4VoHJZ1dgUeJJI5odabHZ7eZITqHQy4PwvMhvZAAJjJgQo3HWAGEIOwUSgoNBISBEnUACpXVzExZwBQKIA)
```sh
--parser babylon
```

**Input:**
```jsx
const tag = 'div';
const template = html`
<${tag}>
</${tag}>
`;

```

**Output:**
```jsx
SyntaxError: Unexpected closing tag "PRETTIER_HTML_PLACEHOLDER_1_IN_JS". It may happen when the tag has already been closed by another tag. For more info see https://www.w3.org/TR/html5/syntax.html#closing-elements-that-have-implied-end-tags (3:1)
  1 | 
  2 | <PRETTIER_HTML_PLACEHOLDER_0_IN_JS>
> 3 | </PRETTIER_HTML_PLACEHOLDER_1_IN_JS>
    | ^
  4 | 
```

**Expected behavior:**

**Comment by kachkaev** [comment]
Perhaps, prettier's placeholder names should be some hash function of normalised ASTs inside `${}`?

```
hash(tag) → ABCDEF

↓

<PRETTIER_HTML_PLACEHOLDER_ABCDEF_IN_JS>
</PRETTIER_HTML_PLACEHOLDER_ABCDEF_IN_JS>
```

**Comment by j-f1** [comment]
Base64 of the normalized AST?

**Comment by kachkaev** [comment]
base64 produces `+-=` in addition to `0-9A-Za-z`, which can break the html parser. This can be a trimmed version of some cheap hash.

We can look into [object-hash](https://www.npmjs.com/package/object-hash), [hash.js](https://www.npmjs.com/package/hash.js) or a like if there’s no hashing algorithm in prettier yet. Just need to make sure we don’t accidentally bloat the bundle with hundreds of crypto functions along with just one of the we need :–D

**Comment by thorn0** [comment]
Can the HTML formatting process ever change the order of the placeholders? If not, can't all the placeholders simply be the same string?

**Comment by ExE-Boss** [comment]
> Can the HTML formatting process ever change the order of the placeholders?

Maybe if they’re in attributes.

---

That said, the same placeholder expression should map to the same `PRETTIER_HTML_PLACEHOLDER_${id}_IN_JS` name.

**Comment by lydell** [comment]
> Possibly if they’re in attributes?

No. Prettier generally never changes order of things. (We don’t do sorting.)

**Comment by thorn0** [comment]
@duailibe Please consider merging #6038 before working on this. That PR is merely a bug fix, it doesn't add any new behavior.

**Comment by Ruben-E** [comment]
Any updates on this? Still seems to be an issue.
